src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
+{
+    /// <summary>
+    /// Not intended for public consumption.
+    /// </summary>
+    public class OutOfProcOrchestrationShim
+    {
+        private readonly IDurableOrchestrationContext context;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OutOfProcOrchestrationShim"/> class.
+        /// </summary>
+        /// <param name="context">The orchestration execution context.</param>
+        public OutOfProcOrchestrationShim(IDurableOrchestrationContext context)
+        {
+            this.context = context ?? throw new ArgumentNullException(nameof(context));
+        }
+
+        private enum AsyncActionType
+        {
+            CallActivity = 0,
+            CallActivityWithRetry = 1,
+            CallSubOrchestrator = 2,
+            CallSubOrchestratorWithRetry = 3,
+            ContinueAsNew = 4,
+            CreateTimer = 5,
+            WaitForExternalEvent = 6,
+            CallEntity = 7,
+            CallHttp = 8,
+        }
+
+        internal async Task HandleOutOfProcExecutionAsync(JObject executionJson)
+        {
+            bool moreWorkToDo = await this.ExecuteAsync(executionJson);
+            if (moreWorkToDo)
+            {
+                // We must delay indefinitely to prevent the orchestration instance from completing.
+                // This is effectively what the Durable Task Framework dispatcher does for normal
+                // orchestration execution.
+                await Task.Delay(Timeout.Infinite);
+            }
+        }
+
+        /// <summary>
+        /// Not intended for public consumption.
+        /// </summary>
+        /// <param name="executionJson">The result of the out-of-proc execution.</param>
+        /// <returns><c>true</c> if there are more executions to process; <c>false</c> otherwise.</returns>
+        public async Task<bool> ExecuteAsync(JObject executionJson)
```

Comment:
Just so that it can be called from unit tests. I suppose we could do an `InternalsVisibleTo` thing with the test project, but I usually try to avoid that because it might cause us to take too many dependencies on internal APIs from our test code...